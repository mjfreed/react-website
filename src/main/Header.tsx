import React, { useEffect, useState } from "react";
import { Dropdown, DropdownItemProps, Menu, MenuItemProps } from "semantic-ui-react";
import { useHistory } from "react-router";
import { MenuLocation } from "../util/constants";
import "../App.css";

// TODO: Fix bug where pointer resets to home on refresh but history does not
// Maybe figure out the dropdown on hover thing

const Header: React.FC = () => {
    const history = useHistory();
    const [pointer, setPointer] = useState("home");

    const handleMenuClick = (event: React.SyntheticEvent, data: MenuItemProps | DropdownItemProps) => {
        const name = data.name ? data.name : "home";
        setPointer(name);
        history.push(MenuLocation[name]);
    };

    return (
        <div>
            <Menu fixed="top" pointing secondary>
                <Menu.Item active={pointer === "home"} as="a" name="home" text="Home" onClick={handleMenuClick} />
                <Dropdown item text="About">
                    <Dropdown.Menu>
                        <Dropdown.Item
                            active={pointer === "aboutMe"}
                            as="a"
                            name="aboutMe"
                            text="Me"
                            onClick={handleMenuClick}
                        />
                        <Dropdown.Item
                            active={pointer === "aboutWebsite"}
                            as="a"
                            name="aboutWebsite"
                            text="Website"
                            onClick={handleMenuClick}
                        />
                    </Dropdown.Menu>
                </Dropdown>
                <Menu.Item
                    active={pointer === "contact"}
                    as="a"
                    name="contact"
                    text="Contact"
                    onClick={handleMenuClick}
                />
            </Menu>
        </div>
    );
};

export default Header;
