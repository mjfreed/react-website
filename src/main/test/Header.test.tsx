import { fireEvent, render } from "@testing-library/react";
import React from "react";
import { act } from "react-dom/test-utils";
import { MemoryRouter } from "react-router";
import Header from "../Header";

const getRenderedHeader = () => {
    return render(
        <MemoryRouter>
            <Header />
        </MemoryRouter>
    );
};

describe("Header.tsx", () => {
    it("renders without crashing", () => {
        const { container } = getRenderedHeader();
        expect(container).toMatchSnapshot();
    });

    it("contains all header items", () => {
        const { getByTestId } = getRenderedHeader();

        const buttons = [];

        buttons.push(getByTestId("header-home"));
        buttons.push(getByTestId("header-about"));
        buttons.push(getByTestId("header-contact"));

        buttons.forEach((button) => expect(button).toBeTruthy());
    });

    it("shows about dropdown menu when About button is clicked", async () => {
        const { getByTestId, findByTestId } = getRenderedHeader();

        const headerButton = getByTestId("header-about");
        fireEvent.click(headerButton);
        await act(async () => {
            expect(await findByTestId("header-about-me")).toBeTruthy();
        });

        await act(async () => {
            expect(await findByTestId("header-about-website")).toBeTruthy();
        });
    });
});
