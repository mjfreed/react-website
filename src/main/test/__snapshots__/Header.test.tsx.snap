// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header.tsx renders without crashing 1`] = `
<div>
  <div>
    <div
      class="ui massive pointing secondary top fixed menu"
      style="background-color: rgb(255, 255, 255);"
    >
      <a
        class="active item"
        data-testid="header-home"
        text="Home"
      >
        Home
      </a>
      <div
        aria-expanded="false"
        class="ui item dropdown"
        data-testid="header-about"
        role="listbox"
        tabindex="0"
      >
        <div
          aria-atomic="true"
          aria-live="polite"
          class="divider text"
          role="alert"
        >
          About
        </div>
        <i
          aria-hidden="true"
          class="dropdown icon"
        />
        <div
          class="menu transition"
        >
          <a
            aria-checked="false"
            class="item"
            data-testid="header-about-me"
            name="about"
            role="option"
          >
            <span
              class="text"
            >
              Me
            </span>
          </a>
          <a
            aria-checked="false"
            class="item"
            data-testid="header-about-website"
            name="about:website"
            role="option"
          >
            <span
              class="text"
            >
              Website
            </span>
          </a>
        </div>
      </div>
      <a
        class="item"
        data-testid="header-contact"
        text="Contact"
      >
        Contact
      </a>
    </div>
  </div>
</div>
`;
